webnn/resources/utils.js

```diff
@@ -1,61 +1,308 @@
+  } else if (metricType === 'ATOL') {
+    assert_array_approx_equals(actual, expected, tolerance, description);
+  }
+};
+
+/**
+ * Check computed results with expected data.
+ * @param {String} operationName - An operation name
+ * @param {Object.<String, MLOperand>} namedOutputOperands
+ * @param {Object.<MLNamedArrayBufferViews>} outputs - The resources of required outputs
+ * @param {Object} resources - Resources used for building a graph
+ */
+const checkResults = (operationName, namedOutputOperands, outputs, resources) => {
+  const metricType = Object.keys(PrecisionMetrics[operationName])[0];
+  const tolerance = getPrecisonTolerance(operationName, metricType, resources);
+  const expected = resources.expected;
+  const operandType = expected.type;
+  let outputData;
+  let expectedData;
+  if (Array.isArray(expected)) {
+    // check outputs by split or gru
+    for (let operandName in namedOutputOperands) {
+      outputData = outputs[operandName];
+      expectedData = getExpectedData(expected, operandName);
+      doAssert(operationName, outputData, expectedData, tolerance, operandType, metricType)
+    }
+  } else {
+    const outputName = expected.name ? expected.name : 'outputOperand';
+    outputData = outputs[outputName];
+    expectedData = expected.data;
+    doAssert(operationName, outputData, expectedData, tolerance, operandType, metricType)
+  }
+};
+
+/**
+ * Build a graph.
+ * @param {MLGraphBuilder} builder - A ML graph builder
+ * @param {Object} resources - Resources used for building a graph
+ * @param {Function} buildFunc - A build function for an operation
+ * @returns [namedOperands, inputs, outputs]
+ */
+const buildGraph = (builder, resources, buildFunc) => {
+  const namedOperands = buildFunc(builder, resources);
+  let inputs = {};
+  if (Array.isArray(resources.inputs)) {
+    // the inputs of concat() is a sequence
+    for (let subInput of resources.inputs) {
+      inputs[subInput.name] = new TypedArrayDict[subInput.type](subInput.data);
+    }
+  } else {
+    for (let inputName in resources.inputs) {
+      const subTestByName = resources.inputs[inputName];
+      inputs[inputName] = new TypedArrayDict[subTestByName.type](subTestByName.data);
+    }
+  }
+  let outputs = {};
+  if (Array.isArray(resources.expected)) {
+    // the outputs of split() or gru() is a sequence
+    for (let i = 0; i < resources.expected.length; i++) {
+      const subExpected = resources.expected[i];
+      outputs[subExpected.name] = new TypedArrayDict[subExpected.type](sizeOfShape(subExpected.shape));
+    }
+  } else {
+    const outputName = resources.expected.name ? resources.expected.name : 'outputOperand';
+    // matmul 1D with 1D produces a scalar which doesn't have its shape
+    const shape = resources.expected.shape ? resources.expected.shape : [1];
+    outputs[outputName] = new TypedArrayDict[resources.expected.type](sizeOfShape(shape));
```

Comment:
The `outputName` should be consistent with operand build function, e.g., `buildConcat()`. Consider using `resources.expected.name` and adding the expected name into JSON file for each test case.